drivers: dma: stm32: bdma support for H7

heinwessels opened the pull request:
Implement STM32H7 BDMA driver and integrate into dmamux

Co-authored-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>
Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels commented:
This functionality is currently being verified using `samples/drivers/adc` in combination with https://github.com/zephyrproject-rtos/zephyr/pull/52965, with the following additional changes:

`samples/drivers/adc/src/main.c`
```C
__in_section(shared_sram4, adc3,
	     dma_ds) static int16_t buffer[ARRAY_SIZE(adc_channels)] __aligned(32);
```

`samples/drivers/adc/nucleo_h743zi.overlay`
```devicetree
#include <zephyr/dt-bindings/adc/adc.h>

/ {
	zephyr,user {
		/*
		 * Note: Data in ADC results buffer will not be in the order
		 * that they are defined here, but as channels selection is done
		 * by a bitmask, they will be sorted instead in a way that
		 * lowest channel number result is on lowest buffer address
		 */
		io-channels = <&adc3 1>, <&adc3 2>;
	};
};

&adc3 {
	dmas = < &dmamux2 0 17 0x2C80>;
	dma-names = "dmamux";

	#address-cells = <1>;
	#size-cells = <0>;

	channel@1 {
		reg = <1>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <12>;
	};

	channel@2 {
		reg = <2>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <12>;
	};
};

&bdma1 {
	status="okay";
};

&dmamux2 {
	status="okay";
};
```

FRASTM commented:
Do you also have tests/drivers/dma/loop_transfer and tests/drivers/dma/chan_blen_transfer PASSED  with bdma on any stm32h7 board ?

heinwessels commented:
> Do you also have tests/drivers/dma/loop_transfer and tests/drivers/dma/chan_blen_transfer PASSED with bdma on any stm32h7 board ?

@FRASTM not yet, wasn't sure about them. I will add tests to a seperate commit in this PR as well. Thanks!

heinwessels commented:
> Do you also have tests/drivers/dma/loop_transfer and tests/drivers/dma/chan_blen_transfer PASSED with bdma on any stm32h7 board ?

Tested, and both tests run successfully. However, both already have an DMA test for the `nucleo_h743zi` for `dmamux1`, which means I can't add another test for `dmamux2`. The test only supports one dma. I'm happy to expand the testing to support multiple dma's if I can pack it into this PR (with specific commits of course).

teburd commented on drivers/dma/CMakeLists.txt:
```diff
@@ -7,6 +7,7 @@ zephyr_library_sources_ifdef(CONFIG_DMA_STM32U5	dma_stm32u5.c)
 zephyr_library_sources_ifdef(CONFIG_DMA_STM32		dma_stm32.c)
 zephyr_library_sources_ifdef(CONFIG_DMA_STM32_V1	dma_stm32_v1.c)
 zephyr_library_sources_ifdef(CONFIG_DMA_STM32_V2	dma_stm32_v2.c)
+zephyr_library_sources_ifdef(CONFIG_BDMA_STM32		bdma_stm32.c)
```
DMA_STM32_BDMA as a Kconfig symbol and dma_stm32_bdma.c as a file name would better match the existing convention

teburd reviewed (changes requested):
Please follow the existing filename/kconfig naming convention

heinwessels committed:
tests: drivers: dma: loop_transfer: add bdma test to nucleo_h743zi

Add DMA unit tests for BDMA driver.

Also requires adding additional kconfig options to allocate
the memory buffers in a specific SRAM section, because
the BDMA only has access to SRAM4.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
tests: drivers: dma: chen_blen_transfer: support testing multiple dmas

Some devices contain multiple dmas

which requires multiple tests

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
tests: drivers: dma: chen_blen_transfer: add bdma test to nucleo_h743zi

Add unit tests for BDMA driver.

Also requires adding additional kconfig options to allocate
the memory buffers in a specific SRAM section, because
the BDMA only has access to SRAM4.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels commented:
I've added tests to `dma/loop_transfer` and `dma/chan_blen_transfer` for the `bdma` on `nucleo_h743zi`. This required a change in the test driver because the tests only supported one active dma, while the `nucleo_h743zi` now has two: `dmamux1` and `dmamux2`.

However, this did bring to light a new bug where `dmamux2` doesn't function when `dmamux1` is also active. When `dmamux1` is disabled then all tests pass. So trying to solve this now.

teburd reviewed (changes requested):
An interesting take on doing tests over all the test_dma described in DT, clearly works and I'm fine with it.

I wonder if we could maybe do this more automatically by moving the st,mem2mem to the dma binding as mem2mem, and then using DT macros to iterate over available dma controllers that support mem2mem instead of having to explicitly define them for the test.

Though there would still need to be an overlay probably to set their status to okay... so probably not worth it. Just a passing thought on the matter.

The file name still doesn't match up with the existing files, please rename the bdma_stm32.[c|h] files to match the existing naming scheme, once done, +1 from me.

teburd commented:
> @FRASTM I've added tests to `dma/loop_transfer` and `dma/chan_blen_transfer` for the `bdma` on `nucleo_h743zi`. This required a change in the test driver because the tests only supported one active dma, while the `nucleo_h743zi` now has two: `dmamux1` and `dmamux2`.
> 
> However, this did bring to light a new bug where `dmamux2` doesn't function when `dmamux1` is also active. When `dmamux1` is disabled then all tests pass. So trying to solve this now.

One option might be to specify a dmamux2 test plan with a different .conf and dtsi... but of course that will require a lot of copy pasta. There's a PR to enable setting namespaced Kconfigs that might be helpful in that regard, take a look

https://github.com/zephyrproject-rtos/zephyr/pull/53747

Using that perhaps you could enable a testcase for stm32 for dmamux1 and dmamux2 without having to create new overlays for each option

teburd commented:
This sort of thing may/may not be helpful, take a look
https://github.com/zephyrproject-rtos/zephyr/pull/53725

heinwessels commented on tests/drivers/dma/loop_transfer/src/test_dma_loop.c:
```diff
@@ -334,19 +330,27 @@ static int test_loop_suspend_resume(void)
-}
+FOR_EACH(LOOP_TEST, (), DMA_LIST);
+
+#define LOOP_SUSPED_RESUME_TEST(dma_name)                                                          \
```
Haha oops. Changed to be exactly the test function name, except all capitals.

heinwessels commented:
> An interesting take on doing tests over all the test_dma described in DT, clearly works and I'm fine with it.

> This sort of thing may/may not be helpful, take a look

That looks useful, and it would definitely be nicer if it could automatically determine the number of testable DMAs. However, if it's all the same to you I would prefer to keep it as is for now. If I have time I could refine it possibly in an future PR. :)

teburd commented:
> 

Absolutely, this is not a blocker

heinwessels commented:
Changed two things:
- Completely decoupled `dma v1/2 + dmamux1` and `bdma + dmamux2` meaning each can compile while the other is disabled. This did result in the following weirdness because `dma_stm32.c` shouldn't be compiled when only `BDMA` is enabled
- Added SRAM section to the linker script
- Added a check during channel configuration to ensure that memory addresses are indeed in SRAM4, because that's the only SRAM location the BDMA has access to.

**Remaining issues**
I've done some more thorough testing using `dma/loop_transfer` and there are still some inconsistencies.
- According to the test the BDMA is inconsistent with the data transfer, which I believe is a caching issue. I've added `SCB_InvalidateDCache_by_Addr` which shows most of the data, but not _all_ the data. There are occationally chunks that show all zeroes.
- The dmamux1 doesn't function at all, but this could be because currently the tx/rx arrays are located in SRAM4, which I don't think the DMA1 has access to. Will verify next week.
- The `chan_blen_transfer` isn't passing, but I'm focussing on getting `dma/loop_transfer` functioning first.

carlescufi commented:
@erwango ping

heinwessels commented:
Okay, now all tests are passing for `tests/drivers/dma/loop_transfer` which tests both `dma1/2` and `bdma`. However, there's still un uglyness which can't stay where for every `STM32H7` board the tx/rx arrays are placed in `SRAM4` because that's what BDMA needs. Ideally we should be able to define where the memory should be placed in the DTS, but I'm not sure yet how to achieve that.

erwango commented:
@heinwessels Can you check the compliance issues ?

heinwessels commented:
> @heinwessels Can you check the compliance issues ?

I should've put this PR as draft. Even though I believe the BDMA and DMAMUX2 is feature complete I think the unit tests still need work to fully test and verify it. And seems I have some conflicts to solve now too. Once that is done I will work on the formatting, because code is subject to change :)

heinwessels committed:
tests: drivers: dma: loop_transfer: support testing multiple dmas

Some devices contain multiple dmas

which requires multiple tests

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels commented:
I've fixed unit tests for `dma/loop_transfer` and `dma/chan_blen_transfer`. I had to add a kconfig option to specify where in the sram the buffers will be placed because BDMA only has access to SRAM4. This then also requried manual cache invalidation through `CONFIG_ARCH_CACHE` because it's no longer in the `.nocache` section.

teburd commented:
DMA test changes seem ok to me, driver looks ok to me. Requesting a review from NXP authors to make sure this works for them as it changes some of the test configs.

Will +1 once I've verified this doesn't break tests on Intel hardware.

erwango commented on drivers/dma/dma_stm32_bdma.c:
```diff
@@ -0,0 +1,940 @@
+
+	bdma_stm32_clear_channel_irq(dev, id);
+
+	/* Check potential BDMA override */
+	if (config->linked_channel == STM32_DMA_HAL_OVERRIDE) {
+		/* BDMA channel is overridden by HAL BDMA
+		 * Retain that the channel is busy and proceed to the minimal
+		 * configuration to properly route the IRQ
+		 */
+		channel->busy = true;
+		channel->hal_override = true;
+		channel->bdma_callback = config->dma_callback;
+		channel->user_data = config->user_data;
+		return 0;
+	}
```
Is that a possible case ?
Today `STM32_DMA_HAL_OVERRIDE` happens when QSPI/OSPI/SDMMC are enabling DMA.
But in those cases, they're using DMA channels not BDMA ones

heinwessels replied:
True. I removed this `if`.

erwango commented on drivers/dma/dma_stm32_bdma.c:
```diff
@@ -0,0 +1,940 @@
+/*
+ * Copyright (c) 2022 Jeroen van Dooren, Nobleo Technology
```
2023, apply everywhere

heinwessels replied:
Updated.

heinwessels commented on drivers/dma/dma_stm32_v1.c:
```diff
@@ -5,7 +5,7 @@
 
 /**
- * @brief DMA low level driver implementation for F2/F4/F7 series SoCs.
+ * @brief DMA low level driver implementation for F2/F4/F7/H7 series SoCs.
```
I'm not sure what you mean? It's only updating an outdated comment, but I will remove this.

erwango commented on drivers/dma/dma_stm32_bdma.h:
```diff
@@ -0,0 +1,103 @@
+typedef uint32_t (*bdma_stm32_check_flag_func)(BDMA_TypeDef *DMAx);
+
+
+#ifdef CONFIG_DMA_STM32_V1
+bool bdma_stm32_is_dme_active(BDMA_TypeDef *DMAx, uint32_t id);
+void bdma_stm32_clear_dme(BDMA_TypeDef *DMAx, uint32_t id);
+bool bdma_stm32_is_fe_active(BDMA_TypeDef *DMAx, uint32_t id);
+void bdma_stm32_clear_fe(BDMA_TypeDef *DMAx, uint32_t id);
+#endif
```
These headers seem useless. Am I missing something ?

heinwessels replied:
Oops, that's an artefact from an older implementation when the `bdma` and `dma_v1` drivers still shared some code. Removed.

erwango commented on dts/arm/st/h7/stm32h743.dtsi:
```diff
@@ -28,6 +28,10 @@
 			dma-requests= <107>;
 		};
 
+		dmamux2: dmamux@58025800 {
+			dma-requests= <107>;
+		};
```
Would be nice to define for other H7 variants

heinwessels replied:
Added it for the other variants as well which defined `dmamux1`. Also found a mistake in `stm32h7a3` where `dma-requests` should be `107` and not `111`. Fixed it in a different commit. 

[Link to Reference Manual](https://www.st.com/resource/en/reference_manual/rm0455-stm32h7a37b3-and-stm32h7b0-value-line-advanced-armbased-32bit-mcus-stmicroelectronics.pdf)

![image](https://user-images.githubusercontent.com/39875036/221159238-10b45178-dcce-470f-b6a9-efad9f34ca04.png)

erwango commented on tests/drivers/dma/chan_blen_transfer/boards/nucleo_h743zi.conf:
```diff
@@ -1,3 +1,16 @@
 CONFIG_NOCACHE_MEMORY=y
+CONFIG_DMA_LOOP_TRANSFER_SRAM_SECTION=".sram4"
+
+# Because the memory is not placed in the default
+# ".nocache" section the cache needs to be
+# invalidated manually
+CONFIG_CACHE=y
+CONFIG_CACHE_MANAGEMENT=y
+CONFIG_ARCH_CACHE=y
```
You should be able to declare sram4 as `RAM_NOCACHE` using `zephyr,memory-region-mpu` property.
This would simplify handling of buffers on user side.

Question is to know if this is ok to declare sram4 as `RAM_NOCACHE` directly in .dtsi (and affect all current potential users) or should it be done as a configuration (either in .dts or .overlay) and a RAM configuration check in the driver. The former would be easier but the later offer more flexibility to users and is backward compatible (so has my preference).

heinwessels replied:
Oh, didn't know I could do that. That would be much easier than the manual cache handling. Will implement it early next week. And I agree, I would rather have it done as a configuration.

> and a RAM configuration check in the driver.

Do you mean that the driver should fail initialization if `nocache` isn't set of the `sram4` section? That would definitely make it much more user friendly, because cache issues are hard to spot and fix.

erwango replied:
> Do you mean that the driver should fail initialization if nocache isn't set of the sram4 section?

Yep

heinwessels replied:
Okay, I've changed the tests to use `zephyr,memory-region-mpu` and it's indeed much cleaner. 

However, I don't think it's currently possible (or easily implementable) to check if a given address is in a noncachable memory region. Would we consider adding this check required/blocking for now? There's nothing of the sort in the existing DMA drivers, which expects the user to handle all cache maintainence themselves.

erwango replied:
What about checking if dt property is set?

heinwessels replied:
> What about checking if dt property is set?

I thought about that, but it doesn't cover all the possible scenarios. Most unit tests for example use the section `.nocache` which is only defined in the linker, and has no reference (that I know of) in the DTS.

erwango replied:
> Most unit tests for example use the section .nocache which is only defined in the linker, and has no reference (that I know of) in the DTS.

My understanding is that BDMA (in current implementation) can only work with SRAM4, which we expect to be defined with RAM_NOCACHE attribute. We are able to check that this specific condition is set, right ?

heinwessels replied:
Check added. I ended up needing a `strcmp`, but seeing it's a once-off during init I think it would be okay.

erwango commented on drivers/dma/dmamux_stm32.c:
```diff
@@ -273,26 +349,50 @@ static const struct dma_driver_api dma_funcs = {
 #define DEV_DMA2 COND_CODE_1(DT_NODE_HAS_STATUS(DT_NODELABEL(dma2), okay), \
 			     DEVICE_DT_GET(DT_NODELABEL(dma2)), NULL)
 
+#define BDMA_1_BEGIN_DMAMUX_CHANNEL DT_PROP_OR(DT_NODELABEL(bdma1), dma_offset, 0)
```
Nit: Please update upper comment which specifies " valid for series having 1 dmamux and 1 or 2 dmas."

heinwessels replied:
Fixed

erwango commented on drivers/dma/dmamux_stm32.c:
```diff
@@ -273,26 +349,50 @@ static const struct dma_driver_api dma_funcs = {
+#define INIT_DMAMUX2_CHANNEL(x, ...) \
+	{ .dev_dma = DEV_BDMA_BINDING(x), .dma_id = BDMA_CHANNEL(x), }
+
+#if DT_SAME_NODE(DT_DRV_INST(0), DT_NODELABEL(dmamux1))
+#define INIT_INST0_CHANNEL(x, ...) INIT_DMAMUX1_CHANNEL(x, ...)
+#define INIT_INST1_CHANNEL(x, ...) INIT_DMAMUX2_CHANNEL(x, ...)
+#else
+#define INIT_INST0_CHANNEL(x, ...) INIT_DMAMUX2_CHANNEL(x, ...)
+#define INIT_INST1_CHANNEL(x, ...) INIT_DMAMUX1_CHANNEL(x, ...)
+#endif
```
Non blocking note: I'm not quite fund of this specific part, but I see the reason.
One suggestion would be to review the macros defined above to take node id as argument, but that might be tricky.

heinwessels replied:
I agree this is not very clean, but I could not get it to work in a cleaner way. So I would be happy to mark this as op-of-scope for this PR if that's okay with you.

hakehuang commented:
not a blocker, but I prefer to implement like https://github.com/zephyrproject-rtos/zephyr/pull/53184#discussion_r1116957922

hakehuang reviewed (dismissed):
all DMA regression tested on NXP FRDM_K64F platfrom. no issues found

erwango commented on drivers/dma/dma_stm32_bdma.c:
```diff
@@ -0,0 +1,927 @@
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 0);
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 1);
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 2);
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 3);
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 4);
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 5);
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 6);
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+BDMA_STM32_DEFINE_IRQ_HANDLER(0, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
```
What about using `LISTIFY(DT_INST_NUM_IRQS(FOO), ...);` here ?

heinwessels replied:
Implemented.

heinwessels replied:
Also implemented for `dma_stm32.c` in seperate commit.

erwango commented on drivers/dma/dma_stm32_bdma.c:
```diff
@@ -0,0 +1,927 @@
+{
+	ARG_UNUSED(dev);
+
+	BDMA_STM32_IRQ_CONNECT(0, 0);
+	BDMA_STM32_IRQ_CONNECT(0, 1);
+	BDMA_STM32_IRQ_CONNECT(0, 2);
+	BDMA_STM32_IRQ_CONNECT(0, 3);
+	BDMA_STM32_IRQ_CONNECT(0, 4);
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+	BDMA_STM32_IRQ_CONNECT(0, 5);
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+	BDMA_STM32_IRQ_CONNECT(0, 6);
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+	BDMA_STM32_IRQ_CONNECT(0, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
```
Ditto

heinwessels replied:
Implemented.

heinwessels replied:
Also implemented for `dma_stm32.c` in seperate commit.

erwango commented on drivers/dma/dma_stm32_bdma.c:
```diff
@@ -0,0 +1,927 @@
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
+/* Either 5 or 6 or 7 or 8 channels for BDMA across all stm32 series. */
```
Likely a copy/paste

heinwessels replied:
Removed.

erwango reviewed (changes requested):
Some last comments.
Also, my request about splitting commits impacting dmamux still stands

FRASTM commented on drivers/dma/dma_stm32_bdma.h:
```diff
@@ -0,0 +1,103 @@
+typedef uint32_t (*bdma_stm32_check_flag_func)(BDMA_TypeDef *DMAx);
+
+
+#ifdef CONFIG_DMA_STM32_V1
```
Could a DMA of type DBMA also be of type V1 ?

heinwessels replied:
Oops, accidentally erased some of my work. This was already deleted. Fixed and removed..

FRASTM commented on soc/arm/st_stm32/stm32h7/sections.ld:
```diff
@@ -16,3 +16,12 @@ SECTION_DATA_PROLOGUE(eth_stm32,(NOLOAD),)
 #endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(sram4), okay)
+SECTION_PROLOGUE (sram4, ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(sram4))) (NOLOAD),)
```
When adding the section for the SRAM4 to map BDMA buffers, are the clock and power already present to supply the D3 domain ?

heinwessels replied:
Not sure what you mean? Adding this linker section will not change the order in SRAM sections or this driver is initialized. And the first time the BDMA knows about SRAM4 is during `INIT_LEVEL_APPLICATION`, and by then everything is obviously started up as expected.

heinwessels committed:
drivers: dma: stm32: bdma support for H7

Implement STM32H7 BDMA driver.

Co-authored-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>
Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
dts: arm: stm32h6a3: fix incorrect dmamux dma-requests

The dma-requests specified for dmamux is changed to
the correct number of 107. This can be found in the
Reference Manual RM0455 Section 17.1.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
drivers: dma: stm32: dmamux: abstact calls to dma drivers

Wrap calls to DMA drivers through fops to allow
different dmamuxes to point to different types of
dma with different function calls.

In preperation to add support for BDMA and DMAMUX2.

Co-authored-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>
Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
drivers: dma: stm32: dmamux: support dmamux2 and bdma

Extends dmamux driver to support DMAMUX 2,
which supports the BDMA on STM32H7 devices.

Co-authored-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>
Signed-off-by: Hein Wessels <heinwessels93@gmail.com>
